src/solver/diagonal/forward_diag.cpp

```diff
@@ -0,0 +1,171 @@
+#include <miopen/diagonal/diag/invoke_params.hpp>
+#include <miopen/diagonal/solvers.hpp>
+#include <miopen/diagonal.hpp>
+#include <miopen/target_properties.hpp>
+
+#define LOCAL_SIZE 256
+
+namespace miopen {
+
+namespace solver {
+
+namespace diagonal {
+
+namespace diag {
+
+bool DiagForward::IsApplicable(const ExecutionContext& context,
+                               const miopen::diagonal::diag::FwdProblemDescription& problem) const
+{
+    std::ignore    = context;
+    auto inputDims = problem.GetInputDesc().GetLengths();
+
+    if(!problem.IsSameType())
+        return false;
+    if(!problem.IsAllPacked())
+        return false;
+    return true;
+}
+
+ConvSolution
+DiagForward::GetSolution(const ExecutionContext& context,
+                         const miopen::diagonal::diag::FwdProblemDescription& problem) const
+{
+    std::ignore = context;
+
+    auto result = ConvSolution{miopenStatusSuccess};
+
+    auto dtype        = problem.GetInputDesc().GetType();
+    auto input_dtype  = miopen::GetDataType(problem.GetInputDesc().GetType());
+    auto output_dtype = miopen::GetDataType(problem.GetOutputDesc().GetType());
+    auto kernel       = KernelInfo{};
+
+    const auto build_params = KernelBuildParameters{
+        {"MIOPEN_USE_FP16", static_cast<int>(dtype == miopenHalf)},
+        {"MIOPEN_USE_FP32", static_cast<int>(dtype == miopenFloat)},
+        {"MIOPEN_USE_FP64", static_cast<int>(dtype == miopenDouble)},
+        {"MIOPEN_USE_BFP16", static_cast<int>(dtype == miopenBFloat16)},
+        {"INPUT_TYPE", input_dtype == "bfloat16" ? "ushort" : input_dtype},
+        {"OUTPUT_TYPE", output_dtype == "bfloat16" ? "ushort" : output_dtype}};
+
+    kernel.comp_options = build_params.GenerateFor(kbp::HIP{});
+
+    if(problem.GetInputDesc().GetLengths().size() == 1)
+    {
+        auto input_numel = problem.GetInputDesc().GetElementSize();
+
+        size_t xlocalsize = LOCAL_SIZE;
+        size_t xgridsize  = AlignUp(input_numel, xlocalsize);
+        size_t ylocalsize = 1;
+        size_t ygridsize  = 1;
+        size_t zlocalsize = 1;
+        size_t zgridsize  = 1;
+
+        kernel.kernel_file = "MIOpenDiag.cpp";
+        kernel.kernel_name = "Diag1dForward";
+
+        kernel.l_wk.push_back(xlocalsize);
+        kernel.l_wk.push_back(ylocalsize);
+        kernel.l_wk.push_back(zlocalsize);
+
+        kernel.g_wk.push_back(xgridsize);
+        kernel.g_wk.push_back(ygridsize);
+        kernel.g_wk.push_back(zgridsize);
+
+        result.construction_params.push_back(kernel);
+
+        result.invoker_factory = [](const std::vector<Kernel>& kernels) {
+            return [=](const Handle& handle_, const AnyInvokeParams& raw_params) {
+                decltype(auto) kernel = handle_.Run(kernels.front());
+                decltype(auto) params =
+                    raw_params.CastTo<miopen::diagonal::diag::FwdInvokeParams>();
+                auto input_numel = params.inputDesc->GetElementSize();
```

Comment:
In lambda func, you can get variable in the outer scope. Remove this line and get the input_numel in the outer scope: `[=](const std::vector<Kernel>& kernels)`

Reply:
solved